flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/JoinOperationFactory.java

```diff
@@ -99,9 +93,8 @@ private void validateCondition(QueryOperation right, JoinType joinType, Resolved
 	}
 
 	private void verifyConditionType(ResolvedExpression condition) {
-		PlannerExpression plannerExpression = expressionBridge.bridge(condition);
-		TypeInformation<?> conditionType = plannerExpression.resultType();
-		if (conditionType != Types.BOOLEAN) {
+		DataType conditionType = condition.getOutputDataType();
+		if (!LogicalTypeChecks.hasRoot(conditionType.getLogicalType(), LogicalTypeRoot.BOOLEAN)) {
```

Comment:
nit: In general I would recommend to store the logical type in a local variable and let the condition look like

```if (hasRoot(conditionType, BOOLEAN))```

for better readability.